semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala

```diff
@@ -166,12 +166,23 @@ object Javacp {
             method.signature.params
           }
 
-        val parameters: List[s.SymbolInformation] = params.zipWithIndex.map {
-          case (param: JavaTypeSignature, i) =>
-            val paramDisplayName = {
-              if (method.node.parameters == null) "param" + i
-              else method.node.parameters.get(i).name
-            }
+        /* Drop the constructor argument that holds the reference to the outer class.
+         * Previous method doesn't work for JDK 21.
+         */
+        val paramsFiltered = params.zipWithIndex.flatMap { case (param, i) =>
+          val paramDisplayName =
+            Option(method.node.parameters)
+              .flatMap(x => Option(x.get(i)))
+              .flatMap(x => Option(x.name))
+              .getOrElse("param" + i)
```

Comment:
`isOuterClassReference` only applies for fields, not params, so I am afraid the suggested code won't work

Reply:
@bjaglin since i am proposing to add a `def isOuterClassReference(name: String)` method, purely by signature it would compile. are you saying that in the context of a parameter, `this$` does not semantically constitute an outer class reference? what is it, then?

Reply:
The current situation is that there are no fields for JDK 21, so we never drop anything while the constructor actually has a reference to the enclosing class as the first parameter. I couldn't figure out how to do it the correct way, but checking for the name $this might not be a great idea coming to think of it. I might need to dig deeper, but will not have time until at least Thursday.

Reply:
> are you saying that in the context of a parameter, `this$` does not semantically constitute an outer class reference? what is it, then?

My understanding is that the name is only present when javac is invoked with `-parameters`, no matter the JDK version (missing for <21, null for >=21), so we can't rely on it. That's why the current code looks for the presence of a field and strips the first parameter if we find a constructor.

In JDK21, we could use the `synthetic` flag instead (since `MethodParameters` is always present, although without names if `-parameters` is not passed), but that wouldn't work for previous versions.

```scala
$ cat Outer.java 
class Outer {
  private class Inner {
   public Inner(Integer i) {
   }
  }
}
```

```
brice@t14:~$ eval $(cs java --jvm 17 --env)
brice@t14:~$ javac Outer.java  && javap -v Outer.Inner | grep "public Outer\$Inner(Outer, java.lang.Integer);" -A 100
  public Outer$Inner(Outer, java.lang.Integer);
    descriptor: (LOuter;Ljava/lang/Integer;)V
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=2, locals=3, args_size=3
         0: aload_0
         1: aload_1
         2: putfield      #1                  // Field this$0:LOuter;
         5: aload_0
         6: invokespecial #7                  // Method java/lang/Object."<init>":()V
         9: return
      LineNumberTable:
        line 3: 0
        line 4: 9
}
SourceFile: "Outer.java"
NestHost: class Outer
brice@t14:~$ javac -parameters Outer.java  && javap -v Outer.Inner | grep "public Outer\$Inner(Outer, java.lang.Integer);" -A 100
  public Outer$Inner(Outer, java.lang.Integer);
    descriptor: (LOuter;Ljava/lang/Integer;)V
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=2, locals=3, args_size=3
         0: aload_0
         1: aload_1
         2: putfield      #1                  // Field this$0:LOuter;
         5: aload_0
         6: invokespecial #7                  // Method java/lang/Object."<init>":()V
         9: return
      LineNumberTable:
        line 3: 0
        line 4: 9
    MethodParameters:
      Name                           Flags
      this$0                         final synthetic
      i
}
SourceFile: "Outer.java"
NestHost: class Outer
```

```
brice@t14:~$ eval $(cs java --jvm 21 --env)
brice@t14:~$ javac Outer.java  && javap -v Outer.Inner | grep "public Outer\$Inner(Outer, java.lang.Integer);" -A 100
  public Outer$Inner(Outer, java.lang.Integer);
    descriptor: (LOuter;Ljava/lang/Integer;)V
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=1, locals=3, args_size=3
         0: aload_0
         1: invokespecial #1                  // Method java/lang/Object."<init>":()V
         4: return
      LineNumberTable:
        line 3: 0
        line 4: 4
    MethodParameters:
      Name                           Flags
      <no name>                      final synthetic
      <no name>
}
SourceFile: "Outer.java"
NestHost: class Outer
brice@t14:~$ javac -parameters Outer.java  && javap -v Outer.Inner | grep "public Outer\$Inner(Outer, java.lang.Integer);" -A 100
  public Outer$Inner(Outer, java.lang.Integer);
    descriptor: (LOuter;Ljava/lang/Integer;)V
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=1, locals=3, args_size=3
         0: aload_0
         1: invokespecial #1                  // Method java/lang/Object."<init>":()V
         4: return
      LineNumberTable:
        line 3: 0
        line 4: 4
    MethodParameters:
      Name                           Flags
      this$0                         final synthetic
      i
}
SourceFile: "Outer.java"
NestHost: class Outer
```

Reply:
> The current situation is that there are no fields for JDK 21, so we never drop anything while the constructor actually has a reference to the enclosing class as the first parameter

Actually, [as commented above](https://github.com/scalameta/scalameta/pull/3385/commits/ec706e976529f32d112f7bdb64ebb1d379afdd15#r1384232597), I believe we do since we don't rely on the presence of `method.node.parameters`, but the presence of a field as heuristic (if we have a `this$` field, then it must be set, so there has to be a synthetic param we don't want to see as first param of all constructors).

I assume both comments about the off-by-one issues when dropping the first param (mine about looking up the name and @kitbellew's about detecting the last element to tag the varargs) are actually unrelated to this PR since both problems exist on master.

We could add an inner class compiled with `-parametes`, with a varargs in a constructor, and add a test asserting that we capture both its name and its varargs nature. I believe it would fail on master and on this PR as-is, but would succeed with [my suggestion](https://github.com/scalameta/scalameta/pull/3385/commits/ec706e976529f32d112f7bdb64ebb1d379afdd15#r1384261170).

Reply:
> I might need to dig deeper, but will not have time until at least Thursday.

No worry, I don't think there is a rush on that as JDK21 is out for a while and javacp is not key to scalameta! Happy to take over to test my suggestion and add that test I am suggesting above if you prefer.

Reply:
Thanks! If you have time to take over to test your suggestions that would be great, but no pressure ;)

Reply:
It was definitely much more complex than I thought, due to the fact that JDK21 might optimize away the field used in the existing heuristic (my bad for questionning your comment above) & because implicit params are not always tagged similarly, but here is my shot at it: https://github.com/scalameta/scalameta/pull/3386.